meshmode/array_context.py

```diff
@@ -271,182 +286,222 @@ def get_container_context_recursively(ary) -> Optional["ArrayContext"]:
-
-    def __rmul__(self, arg):
-        return self._binary_op(operator.mul, arg, self)
-
-    def __rtruediv__(self, arg):
-        return self._binary_op(operator.truediv, arg, self)
-
-    def __rpow__(self, arg):
-        return self._binary_op(operator.pow, arg, self)
-
-    def __rmod__(self, arg):
-        return self._binary_op(operator.mod, arg, self)
-
-    def __rdivmod__(self, arg):
-        return self._binary_op(divmod, arg, self)
-
-    def __pos__(self):
-        return self
-
-    def __neg__(self):
-        return self._unary_op(operator.neg, self)
-
-    def __abs__(self):
-        return self._unary_op(operator.abs, self)
-
-    # }}}
-
-    # {{{ comparison
-
-    def __eq__(self, arg):
-        return self._binary_op(self.array_context.np.equal, self, arg)
-
-    def __ne__(self, arg):
-        return self._binary_op(self.array_context.np.not_equal, self, arg)
-
-    def __lt__(self, arg):
-        return self._binary_op(self.array_context.np.less, self, arg)
-
-    def __gt__(self, arg):
-        return self._binary_op(self.array_context.np.greater, self, arg)
-
-    def __le__(self, arg):
-        return self._binary_op(self.array_context.np.less_equal, self, arg)
-
-    def __ge__(self, arg):
-        return self._binary_op(self.array_context.np.greater_equal, self, arg)
-
-    # }}}
-
-    # {{{ logical
-
-    def __and__(self, arg):
-        return self._binary_op(operator.and_, self, arg)
-
-    def __xor__(self, arg):
-        return self._binary_op(operator.xor, self, arg)
-
-    def __or__(self, arg):
-        return self._binary_op(operator.or_, self, arg)
-
-    def __rand__(self, arg):
-        return self._binary_op(operator.and_, arg, self)
-
-    def __rxor__(self, arg):
-        return self._binary_op(operator.xor, arg, self)
-
-    def __ror__(self, arg):
-        return self._binary_op(operator.or_, arg, self)
-
-    # }}}
-
-
-@get_container_context.register(ArrayContainerWithArithmetic)
-def _get_array_container_ac_arith(ary: ArrayContainerWithArithmetic):
-    return ary.array_context
+            strict_arg = ""
+
+        arg1_entry, arg1_container = arg1
+        arg2_entry, arg2_container = arg2
+        return (f"{op_str.format(arg1_entry, arg2_entry)} "
+                f"for {arg1_entry}, {arg2_entry} "
+                f"in zip({arg1_container}, {arg2_container}{strict_arg})")
+
+    elif isinstance(arg1, tuple):
+        arg1_entry, arg1_container = arg1
+        return (f"{op_str.format(arg1_entry, arg2)} "
+                f"for {arg1_entry} in {arg1_container}")
+
+    elif isinstance(arg2, tuple):
+        arg2_entry, arg2_container = arg2
+        return (f"{op_str.format(arg1, arg2_entry)} "
+                f"for {arg2_entry} in {arg2_container}")
+    else:
+        return op_str.format(arg1, arg2)
+
+
+def with_container_arithmetic(
+        bcast_number=True, bcast_obj_array=None,
+        arithmetic=True, bitwise=False, shift=False,
+        eq_comparison=True, rel_comparison=None):
```

Comment:
* Why is `eq_comparison` True by default? I would set it to False, since most of the things that end up in here are floating point.
* `rel_comparison` I can see being True by default, for things like `np.where`?

Reply:
And also error if `rel_comparison` is true but `eq_comparison` is false?

Reply:
How's 027de9ef1a52ab3f3a77f610de7ea23f38ba791d? I've tried to go for a cautious approach.

Reply:
> And also error if `rel_comparison` is true but `eq_comparison` is false?

IMO: nah. Equality comparison is kind of special in Python (hashing and such), so I *could* see a use case for someone wanting to leave it out and provide their own implementation. For example, Think `numpy`-style semantics vs something that's a bit more compatible with being in a `dict`.

Reply:
Ah, I see. But can't you always fake equality with `x >= y and x <= y`? Won't that cause problems somewhere?